### ISSUE TITLE
tf.keras?

### ISSUE THREAD

**Issue by danaugrs** [bethgelab/foolbox#320]
Is there a way to use this library with the tensorflow keras API `tensorflow.keras`?

**Comment by oGkgN** [comment]
Please have a look at issue #164 and tell us, if that suits your needs.

**Comment by danaugrs** [comment]
Great, will try that - thanks!

**Comment by danaugrs** [comment]
I can't get foolbox to generate any adversarial images.

Model creation:
```python
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

(x_train_raw, y_train_raw), (x_test_raw, y_test_raw) = mnist.load_data()
x_train = x_train_raw.reshape(60000, 784)
x_test = x_test_raw.reshape(10000, 784)
y_train = to_categorical(y_train_raw)
y_test = to_categorical(y_test_raw)

basic_net = Sequential([
    Dense(64, activation='relu', input_shape=(784,)),
    Dense(10, activation='softmax')
])
basic_net.compile(optimizer=Adam(lr=0.0002), metrics=['accuracy'], loss='categorical_crossentropy')
basic_net.fit(x=x_train, y=y_train, epochs=5)
```

Adversarial image generation:
```python
import foolbox
from foolbox.criteria import TargetClass

fool_basic = foolbox.models.TensorFlowModel.from_keras(basic_net, bounds=(0, 255))
criterion = TargetClass(9)

# None of the attacks below work
#attack = foolbox.attacks.GradientAttack(fool_basic, criterion)       # Doesn't find adversarial image (returns None)
#attack = foolbox.attacks.GradientAttack(fool_basic)                  # Doesn't find adversarial image (returns None)
#attack = foolbox.attacks.GradientSignAttack(fool_basic, criterion)   # Doesn't find adversarial image (returns None)
#attack = foolbox.attacks.GradientSignAttack(fool_basic)              # Doesn't find adversarial image (returns None)
#attack = foolbox.attacks.LBFGSAttack(fool_basic, criterion)          # Doesn't find adversarial image (returns None)
#attack = foolbox.attacks.LBFGSAttack(fool_basic)                     # Doesn't find adversarial image (returns None)
#attack = foolbox.attacks.BoundaryAttack(fool_basic, criterion)       # Crashes
#attack = foolbox.attacks.BoundaryAttack(fool_basic)                  # Crashes
#attack = foolbox.attacks.BasicIterativeMethod(fool_basic, criterion) # Doesn't find adversarial image (returns None)
#attack = foolbox.attacks.BasicIterativeMethod(fool_basic)            # Doesn't find adversarial image (returns None)

adversarial = attack(x_train[5], 2) # basic_net predicts x_train[5] to be label 2 (i.e. [0., 0., 1., 0., 0., 0., 0., 0., 0., 0.])
```

What am I doing wrong?

**Comment by oGkgN** [comment]
You are not doing wrong. I confirm that your code in the current state does not yield adversarial examples – but it does not seem to be a faulty use of TensorFlow's or FoolBox's API. I assume that your network, especially the usage of ReLUs, leads to FoolBox failing to attack (maybe some problems to calculate gradients?). However, when changing the activation function of your network to, e.g., `'sigmoid'`, it works!

Here is may slightly modified version of your code:

```Python
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam

import foolbox
import numpy as np
```

```Python
def prepare_mnist():
    (x_train_raw, y_train_raw), (x_test_raw, y_test_raw) = mnist.load_data()
    x_train = x_train_raw.reshape(60000, 784)
    x_test = x_test_raw.reshape(10000, 784)
    y_train = to_categorical(y_train_raw)
    y_test = to_categorical(y_test_raw)
    return (x_train, y_train), (x_test, y_test)

def train_basic_net(x_train, y_train):
    basic_net = Sequential([
        Dense(64, activation='sigmoid', input_shape=(784,)), # <--- Here is the most important change!
        Dense(10, activation='softmax')
    ])
    basic_net.compile(optimizer=Adam(lr=0.0002), metrics=['accuracy'], loss='categorical_crossentropy')
    basic_net.fit(x = x_train, y = y_train, epochs=5)
    return basic_net

def predict_single_label(model, img):
    return np.argmax(model.predict(np.array([img])), axis = 1)[0]

def generate_adv_image(tfkeras_model, image, image_label, target_label = None):
    foolbox_model = foolbox.models.TensorFlowModel.from_keras(tfkeras_model, bounds=(0, 255))

    if target_label is None:
        attack = foolbox.attacks.GradientSignAttack(foolbox_model)
    else:
        criterion = foolbox.criteria.TargetClass(target_label)
        attack = foolbox.attacks.GradientSignAttack(foolbox_model, criterion)
    return attack(image, image_label)
```


```Python
(x_train, y_train), _ = prepare_mnist()
basic_net = train_basic_net(x_train, y_train)
```


```Python
for i in range(10):
    adv_img = generate_adv_image(basic_net, x_train[i], predict_single_label(basic_net, x_train[i]))
    print(adv_img is not None)
```

**Comment by danaugrs** [comment]
I see - thank you!
